database/multistmt/parse.go

```diff
@@ -58,21 +44,71 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, replacement
 	// completed statements, contents of accum will be dumped in here
 	stmts := make([][]byte, 0, 1000)
 
+	tmp := make([]byte, 0, 10)
+	a := 0
 	for err == nil {
+		// this is non-performant but very clear that the buf variable is wiped out
+		//completely between loop iterations
+		buf = make([]byte, ParseBufSize)
 		n, err := reader.Read(buf)
+		trace("tmp(2): '%s', buf: %s, discard: %v\n", tmp, buf, discard)
+		// tmp is the carry-over buffer,
+		//if the previous loop iteration had two few characters to make comparisions,
+		//tmp will have the characters at the point the loop iteration was abandoned(
+		//break'd out of)
+		if len(tmp) > 0 {
+			trace("copying '%s' to buf\n", tmp)
+			buf = append(tmp, buf[:n]...)
+			trace("buf: %s\n", buf)
+			// n needs to include the length of tmp since it was copied into n,
+			// but n originally only held the number of chars read from the
+			// reader.Read(buf) call.
+			n = n + len(tmp)
```

Comment:
instead of maintaining `n` separately, could we just truncate buf to however many bytes were read?  e.g., above just do this:

```
buf = buf[:n]
```

Reply:
we could but that's another allocation, possibly two more because need to expand the buf again on the next loop iteration, not that it probably matters. But this code path is run many times, unfortunately we don't cache the migration bytes.